ddtrace/tracer/sampler.go

```diff
@@ -147,3 +154,273 @@ func (ps *prioritySampler) apply(spn *span) {
+		exactService: service,
+		Rate:         rate,
+	}
+}
+
+// NameRule returns a SamplingRule that applies the provided sampling rate
+// to spans that match the operation name provided.
+func NameRule(name string, rate float64) SamplingRule {
+	return SamplingRule{
+		exactName: name,
+		Rate:      rate,
+	}
+}
+
+// NameServiceRule returns a SamplingRule that applies the provided sampling rate
+// to spans matching both the operation and service names provided.
+func NameServiceRule(name string, service string, rate float64) SamplingRule {
+	return SamplingRule{
+		exactService: service,
+		exactName:    name,
+		Rate:         rate,
+	}
+}
+
+// RateRule returns a SamplingRule that applies the provided sampling rate to all spans.
+func RateRule(rate float64) SamplingRule {
+	return SamplingRule{
+		Rate: rate,
+	}
+}
+
+// match returns true when the span's details match all the expected values in the rule.
+func (sr *SamplingRule) match(s *span) bool {
+	if sr.Service != nil && !sr.Service.MatchString(s.Service) {
+		return false
+	} else if sr.exactService != "" && sr.exactService != s.Service {
+		return false
+	}
+	if sr.Name != nil && !sr.Name.MatchString(s.Name) {
+		return false
+	} else if sr.exactName != "" && sr.exactName != s.Name {
+		return false
+	}
+	return true
+}
+
+// samplingLimiter is used to limit the quantity of spans sampled by the rules sampler
+// and internally maintains measurements for "effective rate".
+type samplingLimiter struct {
+	limiter *rate.Limiter
+
+	// "effective rate" calculations
+	mu            sync.Mutex // guards below fields
+	currentPeriod time.Time  // timestamp, to detect when counters need resetting
+	previousRate  float64    // previous second's rate, averaged with current rate for smoothing
+	allowed       int        // number of spans allowed by rate limiter in the current period
+	seen          int        // number of spans checked by rate limiter in the current period
+}
+
+// samplingDecision returns the rate limiter's decision to allow the span to be sampled, and the
+// effective rate at the time it is called. The effective rate is computed by averaging the rate
+// for the previous second with the current rate
+func (r *samplingLimiter) samplingDecision(now time.Time) (bool, float64) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if d := now.Sub(r.currentPeriod).Truncate(time.Second); d >= time.Second {
+		// enough time has passed to reset the counters
+		if d == time.Second && r.seen > 0 && r.allowed > 0 {
+			// exactly one second, so update previousRate
+			r.previousRate = float64(r.allowed) / float64(r.seen)
+		} else {
+			// more than one second, so reset previous rate
+			r.previousRate = 0.0
+		}
+		r.currentPeriod = now.Truncate(time.Second)
+		r.allowed = 0
+		r.seen = 0
+	}
+
+	r.seen++
+	var sampled bool
+	if r.limiter.AllowN(now, 1) {
+		r.allowed++
+		sampled = true
+	}
+	// calculate effective rate
+	er := (r.previousRate + (float64(r.allowed) / float64(r.seen))) / 2.0
```

Comment:
This algorithm is not correct. When more than one second has passed, or we are at the very first call, `previousRate` is going to be `0`, which means that _nothing_ was allowed to pass through.

Consider this scenario:
* The rate limiter starts and after two seconds 50 out of 100 traces are allowed to pass. The correct rate is then 0.5, but with your algorithm it will be 0.25, which is half of it. As a result, stats will be doubled in the backend. This will happen every time your "reset" happens.

You might have to use `math.NaN` again.

Reply:
After speaking to @furmmon I got:
> To reassure you, this metric main purpose is to know the rate limiter triggered and have an idea it triggers. Today no stats are using it for computation

So you can ignore this comment for now and let's not focus either on this one nor the tests for now.